python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,218 @@
+):
+    """Factory for mixins defining collections of delegated operations.
+
+    This function generates mixins based on two common paradigms in cuDF:
+
+    1. libcudf groups many operations into categories using a common API. These
+       APIs usually accept an enum to delineate the specific operation to
+       perform, e.g. binary operations use the `binary_operator` enum when
+       calling the `binary_operation` function. cuDF Python mimics this
+       structure by having operations within a category delegate to a common
+       internal function (e.g. DataFrame.__add__ calls DataFrame._binaryop).
+    2. Many cuDF classes implement similar operations (e.g. `sum`) via
+       delegation to lower-level APIs before reaching a libcudf C++ function
+       call. As a result, many API function calls actually involve multiple
+       delegations to lower-level APIs that can look essentially identical. An
+       example of such a sequence would be DataFrame.sum -> DataFrame._reduce
+       -> Column.sum -> Column._reduce -> libcudf.
+
+    This factory creates mixins for a category of operations implemented by via
+    this delegator pattern. The resulting mixins make it easy to share common
+    functions across various classes while also providing a common entrypoint
+    for implementing the centralized logic for a given category of operations.
+    Its usage is best demonstrated by example below.
+
+    Parameters
+    ----------
+    mixin_name : str
+        The name of the class. This argument should be the same as the object
+        that this function's output is assigned to, e.g.
+        :code:`Baz = _create_delegating_mixin("Baz", ...)`.
+    docstring : str
+        The documentation string for the mixin class.
+    category_name : str
+        The category of operations for which a mixin is being created. This
+        name will be used to define or access the following attributes as shown
+        in the example below:
+            - f'_{category_name}_DOCSTRINGS'
+            - f'_VALID_{category_name}S'
+            - f'_SUPPORTED_{category_name}S'
+    category_operation_name : str
+        The name given to the core function implementing this category of
+        operations.  The corresponding function is the entrypoint for child
+        classes.
+    supported_ops : List[str]
+        The list of valid operations that subclasses of the resulting mixin may
+        request to be implemented.
+
+    Examples
+    --------
+    >>> MyFoo = _create_delegating_mixin(
+    ...     "MyFoo", "MyFoo's docstring", "FOO", "do_foo", {"foo1", "foo2"}
+    ... )
+
+    >>> # The above is equivalent to defining a class like so:
+    ... class MyFoo:
+    ...     # The set of valid foo operations.
+    ...     _VALID_FOOS = {"foo1", "foo2"}
+    ...
+    ...     # This is the method for child classes to override. Note that the
+    ...     # first parameter is always called "op".
+    ...     def do_foo(self, op, *args, **kwargs):
```

Comment:
```suggestion
    ...     def do_foo(self, op: str, *args, **kwargs):
```
`op` should have an annotation. Otherwise `pop` would fail.

Reply:
I'm going to resolve this comments since I'm going to change it so that annotations aren't required and annotations just clutter examples.